src/engine/loopingcontrol.cpp

```diff
@@ -835,27 +923,104 @@ void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint) {
         } else {
             newloopSamples.end += 2;
         }
-    } else if (newloopSamples.end > samples) {
+    }
+    if (newloopSamples.end > samples) {
         // Do not allow beat loops to go beyond the end of the track
         newloopSamples.end = samples;
```

Comment:
The use case is when the user has accidentally doubled the loop size beyond the end of the track. This is kept track of with a boolean here so slotLoopHalve can know to halve beatloop_size in this case instead of just the loop length. Otherwise, when the user activates loop_halve after doubling to the end of the track, the loop gets halved but it is out of sync with beatloop_size.